// restful-easy/gruntfile.js
'use strict';

module.exports = function(grunt) {
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        watch: {
            js: {
                files: ['<%= jshint.all.src %>'],
                tasks: ['jshint']
            }
        },
        jshint: {
            all: {
                src: ['gruntfile.js', 'server.js', 'test/**/*.js', '!test/coverage/**/*.js'],
                options: {
                    jshintrc: true
                }
            }
        },
        nodemon: {
            test: {
                script: 'test/server.js',
                options: {
                    ext: 'js,json',
                    nodeArgs: ['--debug']
                }
            }
        },
        mochaTest: {
            options: {
                reporter: 'spec'
            },
            unit: {
                src: ['lib/test/*_spec.js']
            },
            integration: {
                src: ['test/integration/**/*_spec.js']
            }
        },
        env: {
            test: {
                NODE_ENV: 'test'
            },
            coverage: {
                APP_DIR_FOR_CODE_COVERAGE: '/test/coverage/instrument/app/'
            }
        },
        clean: {
            coverage: {
                src: ['test/coverage']
            }
        },
        instrument: {
            files: ['lib/**/*.js'],
            options: {
                lazy: true,
                basePath: 'test/coverage/instrument/'
            }
        },
        storeCoverage: {
            options: {
                dir: 'test/coverage/reports'
            }
        },
        makeReport: {
            src: 'test/coverage/reports/**/*.json',
            options: {
                type: 'lcov',
                dir: 'test/coverage/reports',
                print: 'detail'
            }
        },
        jsbeautifier: {
            'default': {
                src: ['<%= jshint.all.src %>', 'package.json'],
                options: {
                    js: {
                        preserveNewlines: true,
                        maxPreserveNewlines: 2
                    }
                }
            },
            'build': {
                src: '<%= jsbeautifier.default.src %>',
                options: {
                    mode: 'VERIFY_ONLY',
                    js: '<%= jsbeautifier.default.options.js%>'
                }
            }
        }
    });

    //Load NPM tasks
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-env');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-istanbul');
    grunt.loadNpmTasks('grunt-mocha-test');
    grunt.loadNpmTasks('grunt-nodemon');
    grunt.loadNpmTasks('grunt-jsbeautifier');

    //Making grunt default to force in order not to break the project.
    grunt.option('force', true);

    // Server task
    grunt.registerTask('server', ['jshint', 'nodemon']);

    // Test task.
    grunt.registerTask('test', function() {
        grunt.option('force', false);
        grunt.task.run('env:test', 'jshint', 'mochaTest');
    });

    grunt.registerTask('coverage', function() {
        grunt.option('force', false);
        grunt.task.run('jshint', 'clean', 'env', 'instrument', 'mochaTest:integration', 'storeCoverage', 'makeReport');
    });

    grunt.registerTask('build', ['coverage', 'jsbeautifier:build']);

    //Default task(s).
    grunt.registerTask('default', ['test', 'jsbeautifier:default']);

};
